localstack/localstack
# Add experimental event ruler

## Motivation

We have several parity gaps and bugs in multiple different implementations for [event pattern matching](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-event-patterns-content-based-filtering.html) (see PR https://github.com/localstack/localstack/pull/10599). We can offer high AWS-parity quickly using the [Amazon EventBridge Event Bus event matching now runs on Open Source Event Ruler v1.5.0](https://aws.amazon.com/about-aws/whats-new/2024/02/amazon-eventbridge-event-bus-matching-event-ruler/) released `event-ruler` Java library from AWS described in [this blog post](https://aws.amazon.com/blogs/opensource/open-sourcing-event-ruler/).


## Changes

* Add an experimental opt-in `EVENT_RULE_ENGINE=java` flag using the Java [event-ruler](https://github.com/aws/event-ruler) library from AWS through [JPype](https://github.com/jpype-project/jpype).
* Add a `MavenPackageInstaller` utility ~~(could probably move to `localstack/packages/core.py` DONE)~~ 

## Testing

1. Set `EVENT_RULE_ENGINE=java`
2. Ensure that `JAVA_HOME` points to a valid JVM installation if you are testing in host mode
3. Uncomment all `SKIP_LABELS` for the test [test_test_event_pattern](https://github.com/localstack/localstack/blob/389d942ac65caf2f824d2ed88fbbe628d1a28b83/tests/aws/services/events/test_event_patterns.py#L87)
4. Run the test `test_test_event_pattern`

### LocalStack Shutdown

@dfangl and @joe4dev tested the LocalStack shutdown as follows:

1. Build LS image using `make docker-build`
2. Start LocalStack using `EVENT_RULE_ENGINE=java DEBUG=1 localstack start`
3. Run a test case for event pattern matching (e.g., `tests.aws.services.events.test_event_patterns.test_test_event_pattern_with_escape_characters) against the running LS instance
4. `docker stop $CONTAINER_ID` should stop the container immediately and not after a 10s delay

=> worked without delay

## Limitations

* The cross-language call from Python to Java is fast but could add a few milliseconds of delay.
* The tight coupling between Python and Java sharing the same process could crash each others language runtime upon unexpected errors.
* We had to disable the JVM shutdown because LocalStack has too many daemon threads running and JPype only triggers the shutdown after all daemon threads terminated to prevent shutting down the JVM while potentially still in use.
* The arm build failed with `ERROR: Could not build wheels for jpype1` due to `gcc: fatal error: cannot execute ‘cc1plus’: execvp: No such file or directory`. Supposedly, JPype [only supports up to Python 3.10](https://stackoverflow.com/a/77674837/6875981) 🤔, which would be a blocker 🔴  (they recently merged a fix for Python 3.12 though). We are using Python 3.11 in the image and jpype1 builds fine on my ARM MacBook using Python 3.11.7.
  * This build error only happened for ARM builds on Linux and required the extra dependency `apt-get install -y g++`.
  * For S3, we had to add this installation step into the make target `docker-run-tests-s3-only` to keep it out of the S3 image.

### Kinesisanalytics

Kinesisanalytics seems incompatible with `EVENT_RULE_ENGINE=java` indicated by the following [failing tests](https://github.com/localstack/localstack-ext/actions/runs/8734151459/job/23964542786):
- `TestKinesisAnalyticsProvider.test_tag_list_tag_untag_resource`
- `TestKinesisAnalyticsProvider.test_input_processing_configuration`
- `TestKinesisAnalyticsProvider.test_application_output`
- `TestKinesisAnalytics.test_run_siddhi_query`

⇒ all these tests keep failing in the AMD64 pipeline. They are not ARM compatible.
⇒ They fail with the assertion: `AssertionError: assert 'STARTING' in ['READY', 'RUNNING']`
⇒ The siddhi query test also fails with a name error `NameError: name 't_siddhi_api_core' is not defined` in the Siddhi loader `PySiddhi/SiddhiLoader.py`

LS uses some badly outdated fork of the unmaintained (5y; LS 3y last commit) PySiddhi project: https://github.com/localstack/PySiddhi. They are using some kind of custom Python⇒Java bridge based on JNI, which could cause interference.

ℹ️ We decided to ignore this limitation for now. If JPype proves successful, it might be a better alternative than PySiddhi ...

## TODO

- [x] Switch default `EVENT_RULE_ENGINE` back from `"java"` (used for CI testing) to `""`
- [x] https://github.com/localstack/docs/pull/1182

## Follow Up

* Fix LS Python-native implementation for `tests.aws.services.lambda_.test_lambda_integration_dynamodbstreams.TestDynamoDBEventSourceMapping.test_dynamodb_event_filter`
  * `numeric_filter` fails due to missing `anything-but` support
  * `exists_false_filter` fails due to broken implementation
  * => It is likely better to adopt Ben's SNS implementation and unify them all because the current status is hardly maintainable.
